.circleci/config.yml

```diff
@@ -1,148 +1,552 @@
+    shell: /bin/bash --login
+    docker:
+      - image: datadog/docker-library:ddtrace_rb_2_2_10
+        environment:
+          - TEST_DATADOG_INTEGRATION: 1
+      - image: postgres:9.6
+        environment:
+          - POSTGRES_PASSWORD=postgres
+          - POSTGRES_USER=postgres
+          - POSTGRES_DB=postgres
+      - image: mysql:5.6
+        environment:
+          - MYSQL_ROOT_PASSWORD=root
+          - MYSQL_PASSWORD=mysql
+          - MYSQL_USER=mysql
+      - image: elasticsearch:2.4
+      - image: redis:3.0
+      - image: mongo:3.5
+      - image: memcached:1.5-alpine
+      - image: datadog/docker-dd-agent
+        environment:
+          - DD_APM_ENABLED=true
+          - DD_BIND_HOST=0.0.0.0
+          - DD_API_KEY=invalid_key_but_this_is_fine
+    steps:
+      - restore_cache:
+          keys:
+            - '{{ .Environment.CIRCLE_CACHE_VERSION }}-bundled-repo-2.2-{{ .Environment.CIRCLE_SHA1 }}'
+      - restore_cache:
+          keys:
+            - '{{ .Environment.CIRCLE_CACHE_VERSION }}-bundle-2.2-{{ checksum ".circleci/bundle_checksum" }}'
+      - run:
+          name: Run tests
+          command: bundle exec rake ci
+  checkout-2.3:
+    working_directory: /app
+    shell: /bin/bash --login
+    docker:
+      - image: datadog/docker-library:ddtrace_rb_2_3_7
+    steps:
+      - checkout
+      - save_cache:
+          key: '{{ .Environment.CIRCLE_CACHE_VERSION }}-repo-2.3-{{ .Environment.CIRCLE_SHA1 }}'
+          paths:
+            - /app
+  build-2.3:
+    working_directory: /app
+    shell: /bin/bash --login
+    docker:
+      - image: datadog/docker-library:ddtrace_rb_2_3_7
+    steps:
+      - restore_cache:
+          keys:
+            - '{{ .Environment.CIRCLE_CACHE_VERSION }}-bundled-repo-2.3-{{ .Environment.CIRCLE_SHA1 }}'
+            - '{{ .Environment.CIRCLE_CACHE_VERSION }}-repo-2.3-{{ .Environment.CIRCLE_SHA1 }}'
+      - run:
+          name: Initialize bundle cache key
+          command: |
+            touch .circleci/bundle_checksum
+      - restore_cache:
+          keys:
+            - '{{ .Environment.CIRCLE_CACHE_VERSION }}-bundle-2.3-{{ checksum ".circleci/bundle_checksum" }}'
+      - run:
+          name: Install gem dependencies
+          command: bundle install
+      - run:
+          name: Delint with Rubocop
+          command: bundle exec rake rubocop
+      - run:
+          name: Install Appraisal gems
+          command: bundle exec appraisal install
+      - run:
+          name: Compute bundle checksum
+          command: |
+            cat Gemfile.lock gemfiles/*.gemfile.lock > .circleci/bundle_checksum
+      - save_cache:
+          key: '{{ .Environment.CIRCLE_CACHE_VERSION }}-bundled-repo-2.3-{{ .Environment.CIRCLE_SHA1 }}'
+          paths:
+            - /app
+      - save_cache:
+          key: '{{ .Environment.CIRCLE_CACHE_VERSION }}-bundle-2.3-{{ checksum ".circleci/bundle_checksum" }}'
+          paths:
+            - /usr/local/bundle
+  test-2.3:
```

Comment:
I wonder if we could do some deduplication of yml code here using [anchors](https://blog.daemonl.com/2016/02/yaml.html).

Similarly how this is handled in this example circle.ci [config](https://github.com/angular/angular/blob/master/.circleci/config.yml#L30)

Reply:
I think this could benefit the `test` step. Won't do the `build` or `checkout` steps much good though, because they're so customized anyways. But maybe we can integrate this in at some level.

Reply:
Thanks! Looks a bit more concise now